client/src/main/java/io/getstream/chat/android/client/utils/observable/ChatEventsObservable.kt

```diff
@@ -0,0 +1,102 @@
+package io.getstream.chat.android.client.utils.observable
+
+import io.getstream.chat.android.client.errors.ChatError
+import io.getstream.chat.android.client.events.ChatEvent
+import io.getstream.chat.android.client.events.ConnectedEvent
+import io.getstream.chat.android.client.events.ConnectingEvent
+import io.getstream.chat.android.client.events.DisconnectedEvent
+import io.getstream.chat.android.client.events.ErrorEvent
+import io.getstream.chat.android.client.models.EventType
+import io.getstream.chat.android.client.socket.ChatSocket
+import io.getstream.chat.android.client.socket.ChatSocketService
+import io.getstream.chat.android.client.socket.SocketListener
+import java.util.Date
+
+internal class ChatEventsObservable(private val socket: ChatSocket) {
+
+    private val subscriptions = mutableSetOf<EventSubscription>()
+    private var eventsMapper = EventsMapper(this)
+
+    private fun onNext(event: ChatEvent) {
+        subscriptions.forEach { subscription ->
+            if (!subscription.isDisposed) {
+                subscription.onNext(event)
+            }
+        }
+        subscriptions.removeAll(Disposable::isDisposed)
```

Comment:
Should this check be here? I think that this logic shouldn't be inside the `onNext` method, this should only care about propagating events.

Reply:
The problem I'm trying to solve here is that I can't remove subscriptions immediately when they're disposed. Used to be that they had a reference to this `Observable`, and asked to be removed, but doing that while the listeners were being iterated was really difficult.

So I opted for cleaning up the unused ones here, after an event's iteration is completed. Not sure if there's a better place to do this.

Reply:
I think that synchronizing the code as @ogkuzmin suggested, could be a way to solve it. But I understand that it may be out of the scope of this PR, I'll resolve this conversation